### DISCUSSION TITLE
How to use MediatR with WPF application?

### DISCUSSION THREAD

**Discussion by sicil1ano** [LuckyPennySoftware/MediatR#735] [Q&A] [answered]
Hi.

I started using MediatR in a new WPF application to send events (implementations of `INotification`) from business logic to view models.
My reasoning is that I don't want to publish events and subscribe to them and then having a tightly coupled implementation, so I thought it might be best to get advantage of the Mediator pattern and hence MediatR.

I'm injecting MediatR using the simple Microsoft DI framework (using ServiceCollection) using something like this:

```
public static IServiceCollection AddMediatR(this IServiceCollection services)
    {
        return services.AddMediatR(Assembly.GetExecutingAssembly())
                       .AddSingleton<IMediatorService, MediatorService>();
    }
```

The executing assembly is where the `IMediatorService`, the `INotification` and `INotificationHandler` implementations live.

```
public class MediatorService : IMediatorService
{
    private readonly IMediator _mediator;

    public MediatorService(IMediator mediator)
    {
        _mediator = mediator;
    }

    public Task NotifyMyNotificationHappenedAsync(MyObject myObject)
    {
        return _mediator.Publish(new MyNotificationHappened{ Property = myObject });
    }
}
```

The IMediatorService gets injected in the business logic. In the business logic, when I need to publish a notification, I use something like this:

` _mediatorService.NotifyMyNotificationHappenedAsync(myObject);`

My VM, inheriting from `ObservableObject` from Microsoft.Toolkit.Mvvm.ComponentModel package, is injected into the DI as transient object and implements `INotificationHandler<MyNotification>`.
The handler should update a Text property when Handle method gets called:

```
public string MyText
    {
        get => _myText;
        set
        {
            _myText = value;
            OnPropertyChanged(nameof(MyText));
        }
    }
```

```
public async Task Handle(MyNotification notification, CancellationToken cancellationToken)
    {
        await _joinableTaskContext.Factory.SwitchToMainThreadAsync();

        MyText += notification.Property.Text;
    }
```

When I run my application and press a button, a window mapped to my VM gets shown.
With it, the business logic publishes the notification and the Handle method gets called. I see that the property gets updated but the UI never updates. I tried to switch to main thread using `JoinableTaskFactory.SwithToMainThreadAsync()` from Microsoft.VisualStudio.Threading package, inside the Handle method, but no update is visible.

I started doubting whether MediatR was publishing the notification to the correct VM instance. 
I added then an instance guid property to my VM, initiliased in the ctor and verified that indeed everytime the Handle method gets called, it gets called on different instances of my VM. I didn't really expect this behaviour, or, at least, I wasn't aware of that.
```

MyVM instance is e7e506ca-d8ad-4928-a36e-e211701be358
.
.
.
MyVM instance is 24962052-82d9-4a9d-b283-95d07f9209dd
```

Am I doing something wrong? Is WPF a valid use-case for MediatR? If yes, how to use it properly?

**Comment by jbogard** [comment, marked as answer]
I'm gonna go with "probably not" because any time I use any XAML-based applications, eventing is something I have to use some framework explicitly for. I was using for example ReactiveUI that correctly hooked up all the UI events. I wouldn't assume MediatR could work across thread contexts.

**Comment by sicil1ano** [comment]
> I'm gonna go with "probably not" because any time I use any XAML-based applications, eventing is something I have to use some framework explicitly for. I was using for example ReactiveUI that correctly hooked up all the UI events. I wouldn't assume MediatR could work across thread contexts.

Hi @jbogard. Sorry for my late reply.

Thanks for your answer. You confirmed what I thought indeed. Sadly, I ended up replacing MediatR, but at least I know now it's not suitable for certain use-cases. Would be this something worth mentioning in the documentation?

**Comment by ghost** [comment]
faced the same issue too, the UI wouldnt update and cannot bind the Inotificationhandler to the Window/Page/UserControl
had to find different way to accomplish this

**Comment by darcome** [comment]
As far as I understand is because Mediatr creates a new instance of the object that implements the INotificationHandler and then executes the Handle method. Therefore if you implement INotificationHandler on the MainWindow (for example) Mediatr creates a new MainWindow and doesn't use the real MainWindow.

Can anyone tell me if this is correct? And if it is, there is a way to change this behaviour?

Thanks in advance

**Reply by lawrence-laz** [reply]
It is correct. A crude work around is to register your object as a singleton, this way both the UI framework and MediatR will get the same instance, ex.:
```csharp
services.AddSingleton<MyViewModel>();
services.AddSingleton<INotificationHandler<MyNotification>>(serviceProvider => serviceProvider.GetRequiredService<MyViewModel>());
```
